Add mid cycle blog for v1.36 release (#54866)
kubernetes/website: content/en/blog/_posts/2026/kubernetes-v1-36-sneak-peak.md

---
layout: blog
title: 'Kubernetes v1.36: Sneak peek'
date: 2026-XX-XX
draft: true
slug: kubernetes-v1-36-sneak-peek
author: >
  Chad Crowell,
  Kirti Goyal,
  Sophia Ugochukwu,
  Swathi Rao,
  Utkarsh Umre
---

Kubernetes v1.36 is coming at the end of April 2026. This release will include removals and deprecations, and it is packed with an impressive number of enhancements. Here are some of the features we are most excited about in this cycle!

Please note that this information reflects the current state of v1.36 development and may change before release.

> ## The Kubernetes API Removal and Deprecation process

Reviewer: ```suggestion
## The Kubernetes API removal and deprecation process 
```

### Deprecation of `.spec.externalIPs` in Service

The `externalIPs` field in Service `spec` is being deprecated, which means you’ll soon lose a quick way to route arbitrary externalIPs to your Services. This field has been a known security headache for years, enabling man-in-the-middle attacks on your cluster traffic, as documented in [CVE-2020-8554](https://github.com/kubernetes/kubernetes/issues/970760). From Kubernetes v1.36 and onwards, you will see deprecation warnings when using it, with full removal planned for v1.43.

If your Services still lean on `externalIPs`, consider using LoadBalancer services for cloud-managed ingress, NodePort for simple port exposure, or Gateway API for a more flexible and secure way to handle external traffic.

You can find more in [KEP-5707: Deprecate service.spec.externalIPs](https://kep.k8s.io/5707)

### Removal of in-tree gitRepo volume driver

The gitRepo volume type has been deprecated since v1.11. Starting Kubernetes v1.36, the `gitRepo` volume plugin is permanently disabled and cannot be turned back on. This change protects clusters from a critical security issue where using `gitRepo` could let an attacker run code as root on the node. 

> Although `gitRepo` has been deprecated for years and better alternatives have been recommended, it was still technically possible to use it in previous releases. From v1.36 onward, that path is closed for good, so any existing workloads depending on `gitRepo` will need to migrate to supported approaches such as init containers or external git-sync style tools.

Author: No, CSI is not a viable direct replacement for gitRepo volumes. GitRepo was for runtime git cloning into pod paths, but CSI drivers focus on persistent storage provisioning, not ephemeral git sync.

Kubernetes v1.36 will include a massive improvement for volume mounting for Pods on SELinux-enforcing systems by swapping slow recursive file relabeling for the efficient `mount -o context=XYZ` option. This applies the correct SELinux label to the entire volume instantly at mount time, slashing Pod startup delays and avoiding out-of-space risks on nearly full disks.

This feature started as beta in v1.28 for ReadWriteOncePod volumes, and gained alpha  metrics and an opt-out option (`PodSecurityContext.SELinuxChangePolicy: Recursive) in v1.32 to catch conflicts, and now will default to all volumes in v1.36 stable, with CSI drivers enabling it via `CSIDriver.Spec.SELinuxMount`. 

However, this feature can lead to breaking changes in v1.37 due to the mixing of privileged and unprivileged pods. Setting `SELinuxChangePolicy` and `SELinux` labels on Pods correctly is the responsibility of the Pod author, whether they are writing a `Deployment`, `StatefulSet`, or `DaemonSet`. Being careless with these settings can lead to a range of problems when Pods share volumes.

To learn more about this enhancement, read the full KEP [here](https://kep.k8s.io/1710)

> ### External signing of service account tokens

Reviewer 2: ```suggestion
### External signing of ServiceAccount tokens
```

Kubernetes now supports external signing of ServiceAccount tokens. This allows clusters to integrate with external key management systems or signing services instead of relying only on internally managed keys.

With this enhancement, the kube-apiserver can delegate token signing to external systems such as cloud key management services or hardware security modules. This improves security and simplifies key management services for clusters that rely on centralized signing infrastructure. 
This feature graduates to stable (GA) in Kubernetes v1.36.

You can find more in [KEP-740: Support external signing of service account tokens](https://kep.k8s.io/740)

### Device taints and tolerations

This enhancement introduces support for taints and tolerations for devices managed through Dynamic Resource Allocation (DRA). It allows administrators to control which workloads are allowed to consume specific hardware devices such as GPUs or accelerators. This improves scheduling control and helps ensure that specialized hardware resources are only used by workloads that explicitly request them

To learn more about this enhancement, read the full KEP here: https://kep.k8s.io/5055

> ### DRA: Add support for partitionable devices

Reviewer 3: ```suggestion
### DRA support for partitionable devices
```
?

For full details, see the [official retirement announcement](/blog/2025/11/11/ingress-nginx-retirement/).

## Featured enhancements of Kubernetes v1.36

The following list of enhancements is likely to be included in the upcoming v1.36 release. This is not a commitment and the release content is subject to change.

### Faster SELinux labelling for volumes (GA) {#volume-selinux-labelling}

Kubernetes v1.36 will include a massive improvement for volume mounting for Pods on SELinux-enforcing systems by swapping slow recursive file relabeling for the
efficient `mount -o context=XYZ` option. This applies the correct SELinux label to the entire volume instantly at mount time, slashing Pod startup delays and
avoiding out-of-space risks on nearly full disks.

This feature was introduced as beta in v1.28 for `ReadWriteOncePod` volumes. In v1.32, it gained metrics and an opt-out option
(`PodSecurityContext.SELinuxChangePolicy: Recursive`) to help catch conflicts. Now in v1.36, it reaches stable and defaults to all volumes, with Pods or
CSIDrivers opting in via `spec.SELinuxMount`.

> However, we beleive this feature is expected to cause breaking changes in v1.37 due to the mixing of privileged and unprivileged pods. Setting `SELinuxChangePolicy` and

Reviewer 3: ```suggestion
However, we expect this feature to create a risk of breaking changes in the future Kubernetes v1.37 realease due to the mixing of privileged and unprivileged pods.
Setting the `seLinuxChangePolicy` field and
```

### External signing of ServiceAccount tokens

As a beta feature, Kubernetes already supports external signing of ServiceAccount tokens. This allows clusters to integrate with external key management systems
or signing services instead of relying only on internally managed keys.

With this enhancement, the kube-apiserver can delegate token signing to external systems such as cloud key management services or hardware security modules. This
improves security and simplifies key management services for clusters that rely on centralized signing infrastructure. 
This feature graduates to stable (GA) in Kubernetes v1.36.

For more details on this enhancement, refer to [KEP-740: Support external signing of service account tokens](https://kep.k8s.io/740)

### DRA Driver support for Device taints and tolerations

Kubernetes v1.33 introduced support for taints and tolerations for physical devices managed through Dynamic Resource Allocation (DRA). Normally, any device can be
used for scheduling. However, this enhancement allows DRA drivers to mark devices as tainted, which ensures that they will not be used for scheduling purposes.

> Alternatively, cluster administrators are create a `DeviceTaintRule` to mark devices that match a certain selection criteria(such as all devices of a certain

Author: This is an API type. In Kubernetes, resource-related APIs and configuration objects typically use PascalCase.

Developers have that responsibility whether they are writing a Deployment, StatefulSet, DaemonSet or even a custom resource that includes a Pod template.
Being careless
with these settings can lead to a range of problems when Pods share volumes.

For more details on this enhancement, refer to  [KEP-1710: Speed up recursive SELinux label change](https://kep.k8s.io/1710)

### External signing of ServiceAccount tokens

As a beta feature, Kubernetes already supports external signing of ServiceAccount tokens. This allows clusters to integrate with external key management systems
or signing services instead of relying only on internally managed keys.

With this enhancement, the kube-apiserver can delegate token signing to external systems such as cloud key management services or hardware security modules. This
improves security and simplifies key management services for clusters that rely on centralized signing infrastructure. 
We expect that this will graduate to stable (GA) in Kubernetes v1.36.

For more details on this enhancement, refer to [KEP-740: Support external signing of service account tokens](https://kep.k8s.io/740)

### DRA Driver support for Device taints and tolerations

> Kubernetes v1.33 introduced support for taints and tolerations for physical devices managed through Dynamic Resource Allocation (DRA). Normally, any device can be

Reviewer: ```suggestion
Kubernetes v1.33 introduced support for taints and tolerations for physical devices managed through [Dynamic Resource Allocation (DRA)](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/). Normally, any device can be
```

used for scheduling. However, this enhancement allows DRA drivers to mark devices as tainted, which ensures that they will not be used for scheduling purposes.
Alternatively, cluster administrators are create a `DeviceTaintRule` to mark devices that match a certain selection criteria(such as all devices of a certain
driver) as tainted. This improves scheduling control and helps ensure that specialized hardware resources are only used by workloads that explicitly request them. 

In Kubernetes v1.36, this feature graduates to beta with more comprehensive testing complete, making it is accessible by default without the need for a feature
flag and open to user feedback. 

> To learn about taints and tolerations, see [taints and tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).  

Reviewer 4: ```suggestion
To learn about taints and tolerations, see [taints and tolerations](/docs/concepts/scheduling-eviction/taint-and-toleration/).  
```

For more details on this enhancement, refer to [KEP-5055: DRA: device taints and tolerations](https://kep.k8s.io/5055).

### DRA support for partitionable devices

Kubernetes v1.36 expands Dynamic Resource Allocation (DRA) by introducing support for partitionable devices, allowing a single hardware accelerator to be split
into multiple logical units that can be shared across workloads.  This is especially useful for high-cost resources like GPUs, where dedicating an entire device
to a single workload can lead to underutilization.

With this enhancement, platform teams can improve overall cluster efficiency by allocating only the required portion of a device to each workload, rather than
reserving it entirely. This makes it easier to run multiple workloads on the same hardware while maintaining isolation and control, helping organizations get more
value out of their infrastructure.

> To learn more about this enhancement, refer to [KEP-4815: DRA Partitionable Drives](https://kep.k8s.io/4815)

Reviewer: ```suggestion
To learn more about this enhancement, refer to [KEP-4815: DRA Partitionable Devices](https://kep.k8s.io/4815)
```

Do you mean devices here?

---
layout: blog
title: 'Kubernetes v1.36: Sneak peek'

> date: 2026-02-25

Reviewer 4: ```suggestion
date: 2026-XX-XX
draft: true
```

### Support external signing of service account tokens

Kubernetes is graduating in-place updates for Pod resources to General Availability (GA). This feature allows users to adjust `cpu` and `memory` resources without restarting Pods or Containers. Previously, such modifications required recreating Pods, which could disrupt workloads, particularly for stateful or batch applications.
Previous Kubernetes releases already allowed you to change infrastructure resources settings (requests and limits) for existing Pods. This allows for smoother [vertical scaling](/docs/concepts/workloads/autoscaling/vertical-pod-autoscale/), improves efficiency, and can also simplify solution development.

The Container Runtime Interface (CRI) has also been improved, extending the `UpdateContainerResources` API for Windows and future runtimes while allowing `ContainerStatus` to report real-time resource configurations. Together, these changes make scaling in Kubernetes faster, more flexible, and disruption-free.
The feature was introduced as alpha in v1.27, graduated to beta in v1.33, and is targeting graduation to stable in v1.35.

> You can find more in [KEP-740: Support external signing of service account tokens](https://github.com/kubernetes/enhancements/issues/740)

Reviewer 4: ```suggestion
You can find more in [KEP-740: Support external signing of service account tokens](https://kep.k8s.io/740)
```
